gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTableMetadataValidatorUtilsTest.java

```diff
@@ -0,0 +1,178 @@
+    // specially from int to long
+    Schema schema5EvolvedFromSchema4 = AvroSchemaUtil.toIceberg(SchemaBuilder.record("schema5")
+        .fields()
+        .requiredLong("field1")
+        .requiredString("field2")
+        .requiredInt("field3")
+        .endRecord());
+    PartitionSpec partitionSpec = PartitionSpec.builderFor(schema5EvolvedFromSchema4)
+        .identity("field1")
+        .build();
+    TableMetadata tableMetadataWithSchema5AndPartitionSpec = TableMetadata.newTableMetadata(schema5EvolvedFromSchema4,
+        partitionSpec, "tableLocationForSchema5WithPartitionSpec", new HashMap<>());
+
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema1AndUnpartitionedSpec,
+        tableMetadataWithSchema5AndPartitionSpec, SCHEMA_MISMATCH_EXCEPTION);
+  }
+
+  @Test
+  public void testValidateSamePartitionSpec() throws IOException {
+    IcebergTableMetadataValidatorUtils.failUnlessCompatibleStructure(
+        tableMetadataWithSchema1AndPartitionSpec1, tableMetadataWithSchema1AndPartitionSpec1,
+        PARTITION_SPEC_STRICT_EQUALITY_CHECK
+    );
+  }
+
+  @Test
+  public void testValidatePartitionSpecWithDiffName() {
+    PartitionSpec partitionSpec12 = PartitionSpec.builderFor(schema1)
+        .identity("field2")
+        .build();
+    TableMetadata tableMetadataWithSchema1AndPartitionSpec12 = TableMetadata.newTableMetadata(schema1, partitionSpec12,
+        "tableLocationForSchema1WithPartitionSpec12", new HashMap<>());
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema1AndPartitionSpec1,
+        tableMetadataWithSchema1AndPartitionSpec12, PARTITION_SPEC_MISMATCH_EXCEPTION);
+  }
+
+  @Test
+  public void testValidatePartitionSpecWithUnpartitioned() {
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema1AndUnpartitionedSpec,
+        tableMetadataWithSchema1AndPartitionSpec1, PARTITION_SPEC_MISMATCH_EXCEPTION);
+  }
+
+  @Test
+  public void testPartitionSpecWithDifferentTransform() {
+    PartitionSpec partitionSpec12 = PartitionSpec.builderFor(schema1)
+        .truncate("field1", 4)
+        .build();
+    TableMetadata tableMetadataWithSchema1AndPartitionSpec12 = TableMetadata.newTableMetadata(schema1, partitionSpec12,
+        "tableLocationForSchema1WithPartitionSpec12", new HashMap<>());
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema1AndPartitionSpec1,
+        tableMetadataWithSchema1AndPartitionSpec12, PARTITION_SPEC_MISMATCH_EXCEPTION);
+  }
+
+  private void verifyFailUnlessCompatibleStructureIOException(TableMetadata tableAMetadata,
+      TableMetadata tableBMetadata, String expectedMessage) {
+    IOException exception = Assert.expectThrows(IOException.class, () -> {
+      IcebergTableMetadataValidatorUtils.failUnlessCompatibleStructure(tableAMetadata, tableBMetadata,
+          PARTITION_SPEC_STRICT_EQUALITY_CHECK);
```

Comment:
did I miss a test where strict equality was false?  I'd be very interested to see test inputs which pass w/ this is false but raise the exception when this is true

Reply:
Added